File: docs/scripts/qoder_cn_checkin.mjs
#!/usr/bin/env node
/** Qoder CN campaign claims. See ../reverse-note/qoder-cn-daily-checkin_CN.md. */
import fs from 'node:fs';
import { pathToFileURL } from 'node:url';

const BASE = 'https://openapi.qoder.com.cn';
const CAMPAIGNS = '/sash/api/v1/me/campaigns';
const HELP = `Usage: node docs/scripts/qoder_cn_checkin.mjs --auth <file> [--claim] [--json]

Requires Node.js 18+; no dependencies.
Default: query only. --claim submits all claimable CLAIM_BENEFIT campaigns.
The credential file must contain access_token; machine_id is optional.
No automatic token refresh, retries, or scheduling. Credentials are never printed.`;

export function parseArgs(args) {
  const options = { claim: false, json: false, help: false };
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--claim') options.claim = true;
    else if (arg === '--json') options.json = true;
    else if (arg === '--help' || arg === '-h') options.help = true;
    else if (arg === '--auth') {
      if (!args[i + 1] || args[i + 1].startsWith('--')) throw new Error('--auth requires a file path');
      options.auth = args[++i];
    } else throw new Error(`Unknown argument: ${arg}`);
  }
  if (!options.help && !options.auth) throw new Error('--auth is required');
  return options;
}

export async function run(credential, options, fetchImpl = fetch) {
  const token = credential.access_token ?? credential.accessToken;
  const machineId = credential.machine_id ?? credential.machineId;
  if (typeof token !== 'string' || !token.trim()) throw new Error('Missing access_token');
  const headers = {
    Accept: 'application/json',
    Authorization: `Bearer ${token.trim()}`,
    'User-Agent': 'Qoder',
    'Cosy-ClientType': '10',
    'Cosy-Business-Product': 'app',
    'Cosy-Version': '0.2.5',
  };
  if (typeof machineId === 'string' && machineId.trim()) headers['Cosy-MachineId'] = machineId.trim();

  async function request(endpoint, method = 'GET') {
    // Refuse redirects so the bearer cannot be forwarded to another endpoint.
    // Do not emit raw response bodies or network errors containing credentials.
    let response;
    try {
      response = await fetchImpl(BASE + endpoint, { method, headers, redirect: 'error' });
    } catch {
      throw new Error('Network request failed; no automatic retry performed');
    }
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    let payload;
    try { payload = await response.json(); } catch { throw new Error('Invalid JSON response'); }
    return payload?.data ?? payload;
  }

  const payload = await request(CAMPAIGNS);
  const campaigns = Array.isArray(payload) ? payload : payload?.campaigns;
  if (!Array.isArray(campaigns)) throw new Error('Invalid campaigns response: expected an array');
  const result = {
    dryRun: !options.claim,
    campaigns: campaigns.filter(c => c && typeof c === 'object').map(c => ({
      campaignId: c.campaignId, campaignKey: c.campaignKey,
      actionType: c.actionType, claimStatus: c.claimStatus, benefit: c.benefit,
    })),
    claimed: [], skipped: [], errors: [],
  };
  const seen = new Set();
  for (const c of result.campaigns) {
    if (c.actionType !== 'CLAIM_BENEFIT' || c.claimStatus !== 'CLAIMABLE') continue;
    if (typeof c.campaignId !== 'string' || !c.campaignId.trim()) {
      result.errors.push({ error: 'Claimable campaign has no valid campaignId' });
      continue;
    }
    if (seen.has(c.campaignId)) continue;
    seen.add(c.campaignId);
    if (!options.claim) { result.skipped.push(c.campaignId); continue; }
    try {
      // The official iframe submits a bodyless POST (not a JSON body).
      const response = await request(`${CAMPAIGNS}/${encodeURIComponent(c.campaignId)}/claim`, 'POST');
      if (response?.status !== 'CLAIMED') throw new Error('Unexpected claim response status');
      result.claimed.push({ campaignId: c.campaignId, campaignKey: c.campaignKey, amount: c.benefit?.amount });
    } catch (error) {
      result.errors.push({ campaignId: c.campaignId, error: error.message });
    }
  }
  return result;
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  if (options.help) { console.log(HELP); return; }
  let credential;
  try { credential = JSON.parse(fs.readFileSync(options.auth, 'utf8')); }
  catch { throw new Error('Cannot read or parse the credential JSON file'); }
  if (!credential || typeof credential !== 'object') throw new Error('Invalid credential object');
  const result = await run(credential, options);
  if (options.json) console.log(JSON.stringify(result, null, 2));
  else {
    console.log(result.dryRun ? 'Query only; no claims submitted.' : 'Claim mode.');
    for (const c of result.campaigns) console.log(`${c.claimStatus ?? '?'}  ${c.actionType ?? '?'}  ${c.campaignKey ?? c.campaignId ?? '?'}  ${c.benefit?.amount ?? ''}`);
    console.log(`Claimed: ${result.claimed.length}; skipped: ${result.skipped.length}; errors: ${result.errors.length}`);
    for (const e of result.errors) console.error(`${e.campaignId ?? 'campaign'}: ${e.error}`);
  }
  if (result.errors.length) process.exitCode = 1;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch(error => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}

File: docs/scripts/qoder_cn_checkin.py
#!/usr/bin/env python3
"""Qoder CN campaign claims; see ../reverse-note/qoder-cn-daily-checkin_CN.md."""

import argparse
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

BASE = "https://openapi.qoder.com.cn"
CAMPAIGNS = "/sash/api/v1/me/campaigns"


class CheckinError(Exception):
    """A sanitized error safe to display without exposing credentials."""


class NoRedirect(urllib.request.HTTPRedirectHandler):
    """Never forward the bearer token through a redirect."""

    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def request_json(endpoint, method, headers):
    url = BASE + endpoint
    parsed = urllib.parse.urlsplit(url)
    if parsed.scheme != "https" or parsed.netloc != "openapi.qoder.com.cn":
        raise CheckinError("Untrusted API URL")
    # Scheme and authority are explicitly allowlisted above; redirects are disabled below.
    request = urllib.request.Request(url, headers=headers, method=method)  # noqa: S310
    opener = urllib.request.build_opener(NoRedirect())
    try:
        with opener.open(request) as response:
            if not 200 <= response.status < 300:
                raise CheckinError(f"HTTP {response.status}")
            try:
                return json.load(response)
            except (ValueError, UnicodeError):
                raise CheckinError("Invalid JSON response") from None
    except urllib.error.HTTPError as error:
        status = error.code
        error.close()
        raise CheckinError(f"HTTP {status}") from None
    except CheckinError:
        raise
    except Exception:
        raise CheckinError("Network request failed; no automatic retry performed") from None


def unwrap(payload):
    if isinstance(payload, dict) and payload.get("data") is not None:
        return payload["data"]
    return payload


def run(credential, claim=False, transport=request_json):
    if not isinstance(credential, dict):
        raise CheckinError("Invalid credential object")
    token = credential.get("access_token", credential.get("accessToken"))
    machine_id = credential.get("machine_id", credential.get("machineId"))
    if not isinstance(token, str) or not token.strip():
        raise CheckinError("Missing access_token")
    headers = {
        "Accept": "application/json",
        "Authorization": f"Bearer {token.strip()}",
        "User-Agent": "Qoder",
        "Cosy-ClientType": "10",
        "Cosy-Business-Product": "app",
        "Cosy-Version": "0.2.5",
    }
    if isinstance(machine_id, str) and machine_id.strip():
        headers["Cosy-MachineId"] = machine_id.strip()

    payload = unwrap(transport(CAMPAIGNS, "GET", headers))
    campaigns = payload if isinstance(payload, list) else (
        payload.get("campaigns") if isinstance(payload, dict) else None
    )
    if not isinstance(campaigns, list):
        raise CheckinError("Invalid campaigns response: expected an array")
    keys = ("campaignId", "campaignKey", "actionType", "claimStatus", "benefit")
    result = {
        "dryRun": not claim,
        "campaigns": [{k: c.get(k) for k in keys} for c in campaigns if isinstance(c, dict)],
        "claimed": [], "skipped": [], "errors": [],
    }
    seen = set()
    for campaign in result["campaigns"]:
        if campaign["actionType"] != "CLAIM_BENEFIT" or campaign["claimStatus"] != "CLAIMABLE":
            continue
        campaign_id = campaign["campaignId"]
        if not isinstance(campaign_id, str) or not campaign_id.strip():
            result["errors"].append({"error": "Claimable campaign has no valid campaignId"})
            continue
        if campaign_id in seen:
            continue
        seen.add(campaign_id)
        if not claim:
            result["skipped"].append(campaign_id)
            continue
        try:
            endpoint = f"{CAMPAIGNS}/{urllib.parse.quote(campaign_id, safe='')}/claim"
            # Match the official iframe: POST without a JSON body.
            response = unwrap(transport(endpoint, "POST", headers))
            if not isinstance(response, dict) or response.get("status") != "CLAIMED":
                raise CheckinError("Unexpected claim response status")
            benefit = campaign.get("benefit")
            result["claimed"].append({
                "campaignId": campaign_id,
                "campaignKey": campaign["campaignKey"],
                "amount": benefit.get("amount") if isinstance(benefit, dict) else None,
            })
        except CheckinError as error:
            result["errors"].append({"campaignId": campaign_id, "error": str(error)})
    return result


def _auth_dir_files(directory, prefix=None):
    if not directory.is_dir():
        raise CheckinError("--auth-dir is not a directory")
    files = sorted(p for p in directory.glob("*.json")
                   if p.is_file() and (prefix is None or p.name.startswith(prefix)))
    if not files:
        raise CheckinError("No matching .json credential files")
    return files


def load_session(path):
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError, UnicodeError):
        raise CheckinError("Cannot read or parse the credential JSON file") from None
    if not isinstance(data, dict):
        raise CheckinError("Invalid credential object")
    # Do not send credentials belonging to other providers to Qoder.
    if data.get("type") not in (None, "qoder-cn"):
        raise CheckinError("Credential type is not qoder-cn")
    return data


def main(argv=None):
    parser = argparse.ArgumentParser(description="Qoder CN daily check-in (default: checkin).")
    parser.add_argument("action", nargs="?", default="checkin", choices=["status", "checkin"])
    source = parser.add_mutually_exclusive_group()
    source.add_argument("--auth-file", "--auth", dest="auth_file", type=Path)
    source.add_argument("--auth-dir", type=Path, help="Process matching .json files sequentially")
    source.add_argument("--token", help="Manual access token (prefer a file to avoid shell history)")
    parser.add_argument("--prefix", help="Filename prefix filter for --auth-dir")
    parser.add_argument("--machine-id", help="Optional machine ID for --token")
    parser.add_argument("--claim", action="store_true", help="Legacy alias for checkin")
    parser.add_argument("--json", action="store_true", help="Output JSON")
    args = parser.parse_args(argv)
    if args.prefix is not None and args.auth_dir is None:
        parser.error("--prefix requires --auth-dir")
    if args.machine_id is not None and args.token is None:
        parser.error("--machine-id requires --token")
    if args.claim and args.action == "status":
        parser.error("status cannot be combined with --claim")
    if args.auth_file is None and args.auth_dir is None and args.token is None:
        env_file = os.environ.get("QODER_CN_AUTH_FILE")
        if not env_file:
            parser.error("Use --auth-file, --auth-dir, --token or QODER_CN_AUTH_FILE")
        args.auth_file = Path(env_file)
    try:
        files = _auth_dir_files(args.auth_dir, args.prefix) if args.auth_dir else [args.auth_file]
    except CheckinError as error:
        print(f"Error: {error}", file=sys.stderr)
        return 1
    entries = []
    failed = 0
    for path in files:
        entry: dict = {"source": str(path) if path else "manual token"}
        try:
            credential = load_session(path) if path else {
                "access_token": args.token, "machine_id": args.machine_id,
            }
            result = run(credential, claim=args.action == "checkin")
            entry["result"] = result
            if result["errors"]:
                failed += 1
        except CheckinError as error:
            entry["error"] = str(error)
            failed += 1
        except Exception:
            entry["error"] = "Operation failed; no automatic retry performed"
            failed += 1
        entries.append(entry)
        if not args.json:
            print(f"[i] Credential: {entry['source']}")
            if "error" in entry:
                print(f"[x] {entry['error']}")
                continue
            result = entry["result"]
            print("Query only." if result["dryRun"] else "Check-in mode.")
            for campaign in result["campaigns"]:
                print(f"  {campaign['claimStatus']}  {campaign['actionType']}  "
                      f"{campaign['campaignKey'] or campaign['campaignId']}")
            print(f"Claimed: {len(result['claimed'])}; skipped: {len(result['skipped'])}; "
                  f"errors: {len(result['errors'])}")
            for error in result["errors"]:
                print(f"[x] {error['error']}")
    summary = {"total": len(entries), "succeeded": len(entries) - failed, "failed": failed}
    if args.json:
        print(json.dumps({"accounts": entries, "summary": summary}, ensure_ascii=False, indent=2))
    else:
        print(f"Total: {summary['total']}; succeeded: {summary['succeeded']}; failed: {failed}")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())
